src/DataSet/ObjectDataSet.php

```diff
@@ -13,106 +13,48 @@
-        $this->cacheKey = get_class($this->object) . '_' . $this->propertyVisibility;
+        $this->rulesProvided = $this->object instanceof RulesProviderInterface;
+        $this->cacheKey = $this->object::class . '_' . $this->propertyVisibility;
     }
 
-    public function getRules(): iterable
+    public function getObject(): object
     {
-        $objectHasRules = $this->object instanceof RulesProviderInterface;
-        $rules = $objectHasRules ? $this->object->getRules() : [];
-
-        // Providing data set assumes object has its own attributes and rules getting logic. So further parsing of
-        // Reflection properties and rules is skipped intentionally.
-        if ($this->dataSetProvided || $objectHasRules === true) {
-            return $rules;
-        }
-
-        if ($this->hasCachedRules()) {
-            return $this->getCachedRules();
-        }
-
-        foreach ($this->getReflectionProperties() as $property) {
-            $attributes = $property->getAttributes(RuleInterface::class, ReflectionAttribute::IS_INSTANCEOF);
-            foreach ($attributes as $attribute) {
-                $rule = $attribute->newInstance();
-                $rules[$property->getName()][] = $rule;
-
-                if ($rule instanceof AttributeEventInterface) {
-                    $rule->afterInitAttribute($this);
-                }
-            }
-        }
-
-        $this->setCachedRules($rules);
-
-        return $rules;
+        return $this->object;
     }
 
-    /**
-     * @return ReflectionProperty[] Used to avoid error "Typed property must not be accessed before initialization".
-     */
-    private function getReflectionProperties(): array
+    public function getRules(): iterable
     {
-        if ($this->reflectionProperties !== null) {
-            return $this->reflectionProperties;
+        if ($this->rulesProvided) {
+            return $this->object->getRules();
         }
 
-        // Providing data set assumes object has its own attributes and rules getting logic. So further parsing of
-        // Reflection properties and rules is skipped intentionally.
         if ($this->dataSetProvided) {
-            $this->reflectionProperties = [];
-
-            return $this->reflectionProperties;
-        }
-
-        // TODO: Use Generator to collect attributes
-
-        $reflection = new ReflectionObject($this->object);
-        $reflectionProperties = [];
-
-        foreach ($reflection->getProperties($this->propertyVisibility) as $property) {
-            if (PHP_VERSION_ID < 80100) {
-                $property->setAccessible(true);
-            }
-
-            $reflectionProperties[$property->getName()] = $property;
+            return [];
         }
 
-        $this->reflectionProperties = $reflectionProperties;
-
-        return $reflectionProperties;
-    }
-
-    public function getObject(): object
-    {
-        return $this->object;
+        return $this->getCachedRules();
```

Comment:
In my opinion here should be this sequence:

1. Attempt to get rules from cache. If they are found - return cached version.
2. Actual rules getting logic.
3. Update cache and return rules.

Similar with reflection properties.

Reply:
If we do this, it will take several passes through the properties (foreach for reflection properties). In the current implementation we use foreach once.

Reply:
Discussed it in PM. Yes, an additional pass is required, but on the cached data. And with the amount of properties the profit will likely be miserable. But separation is more organized and readable ("design first, optimize later" principle).